content/blog/2020-08-05-cml-self-hosted-runners-on-demand-with-gpus.md

```diff
@@ -0,0 +1,183 @@
+[recently-released project](https://dvc.org/blog/cml-release), Continuous
+Machine Learning (CML), our Docker image acts as a thin wrapper over GitLab and
+GitHub runners, adding some extra capabilities.
+
+Here are some benefits of using CML as a self-hosted runner:
+
+1.  Easy to use. Working the same way for both GitLab and GitHub.
+
+2.  Get out of dependency hell. We tend to install packages (on top of packages,
+    on top of packages…) while we‘re experimenting with models. In ML in
+    particular, we can be dependent on drivers AND libraries, and sometimes
+    precise versions of them (CUDA and TensorFlow, anyone?). Your CI workflow
+    will install all the dependencies in the containerised runner leaving your
+    machine always clean.
+
+3.  Security. If your repo is public your runners could be accesed by anyone
+    that could add
+    [scripts that exploits your machine](https://docs.GitHub.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories).
+    With the containerised runner you are restricting the access to your real
+    machine.
+
+4.  Gain reproducibility. One of the biggest technical debts in the ML space is
+    reproducibility. A few weeks post-experiment, we often discover that trying
+    to put your model back in shape is a pain. Looking at our repo, it’s not
+    obvious what data or training infrastructure or dependencies went into a
+    given result. When you move your ML experiments into a CI/CD system you are
+    making a contract of the dependencies and hardware used for your experiment.
+    Having that contract isolated by the containerised runner, your experiment
+    is perfectly reproducible by anyone in the future.
+
+## Hands on GPU Self-hosted runners 101
+
+### 1) Install nvidia drivers and nvidia-docker in your machine (ubuntu 18.04)
+
+```dvc
+$ curl -s -L https://nvidia.GitHub.io/nvidia-docker/gpgkey | sudo apt-key add - && \
+  curl -s -L https://nvidia.GitHub.io/nvidia-docker/ubuntu18.04/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list && \
+  sudo apt update && sudo apt install -y ubuntu-drivers-common  && \
+  sudo ubuntu-drivers autoinstall  && \
+  sudo apt install -y nvidia-container-toolkit && \
+  sudo systemctl restart docker
+```
+
+You can test that your gpus are up and running with the following command:
+
+```dvc
+$ docker run --gpus all dvcorg/cml-py3 nvidia-smi
+```
+
+We should see something like this:
+![](/uploads/images/2020-08-05/nvidia-smi-output.png)
+
+### 2) Start your self-hosted runner
+
+With CML docker images launching your own self-hosted runner is very easy. They
+comes in two flavours, depending if you need python 2 or 3:
+
+- dvcorg/cml:latest
+- dvcorg/cml-py3:latest
+
+These images have CML and DVC preinstalled (among other perks), plus CUDA
+drivers and Python 2.7. The cml-py3 image comes with Python 3.X as well. That's
```

Comment:
curious - why do we support Python 2 still?

Reply:
many projects in ML space only supports python 2

Reply:
hmm ... any stats on this?

Reply:
No

Reply:
does it mean that with the default names like `dvcorg/cml` we actually provide Python 2 image?

Reply:
Not a python version per se. Simply we do not provide one, its de version that comes bundled with ubuntu 18.04 which is the one used in GH

Reply:
ah, I see ... that's really sad to see that things move so slow, even for Ubuntu and GH

Reply:
true

Reply:
> comes with Python 3.X as well

That's weird though. Implies that `cml-py3` bloatedly has both py2 and py3 packages pre-installed. Hopefully that is not the case and this should be reworded.

Reply:
```suggestion
drivers and Python 2.7 (use the cml-py3 image for Python 3.X). That's
```

Reply:
actually that sentence is very redundant. Its being explained above. im removing it